src/basics.md

```diff
@@ -470,6 +471,80 @@ fn main() {
 }
 ```
 
+[ex-phone]: #ex-phone
+<a name="ex-phone"></a>
+## Extract phone numbers from text
+
+Attempts to process a static string as a list including US phone numbers.  The
+expression is checking that there are 7 digits that come in groups of 3 3 4,
+separated by spaces, dots, dashes, and parenthesis.  The phone number is then
+tested based on formatting rules.
+
+```rust
+# #[macro_use]
+# extern crate error_chain;
+extern crate regex;
+
+use regex::Regex;
+use std::fmt;
+
+# error_chain!{
+#     foreign_links {
+#         Regex(regex::Error);
+#         Io(std::io::Error);
+#     }
+# }
+#
+struct PhoneNumber {
+    country: String,
+    number: String,
+}
+
+// This allows printing phone numbers based on country convention.
+impl fmt::Display for PhoneNumber {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.country.as_ref() {
+            "US" => write!(f, "+1 ({}) {}-{}", &self.number[0..3], &self.number[3..6], &self.number[6..10]),
+            _    => write!(f, "{}", self.number),
+        }
+    }
+}
+
+fn run() -> Result<()> {
+    let phone_text = "
+    +1 505 881 9292
+    (202) 991 9534
+    Alex 5553920011
+    1 (800) 233-2010
+    1.299.339.1020";
+    let re = Regex::new(r"(?x)
+                         \+?1?                      #Country Code Optional
+                         [\x20\(\.]*
+                         ([2-9]\d{2})               #Area Code
+                         [\x20\)\.\-]*
+                         ([2-9]\d{2})               #Exchange Code
+                         [\x20\.\-]*
+                         (\d{4})                    #Subscriber Number")?;
+    let mut phone_numbers = re.captures_iter(phone_text).map(|cap| {
+        PhoneNumber {
+            country: "US".to_owned(),
+            number: [cap[1].to_string(),cap[2].to_string(),cap[3].to_string()].concat()
+        }
+    });
+    assert!(format!("{}", phone_numbers.next().unwrap()) == "+1 (505) 881-9292");
```

Comment:
that is why I've suggested something like
assert_eq!(phone_numbers.next().map(String::from), Some("+1 (299) 339-1020".to_owned()));